### ISSUE TITLE
clang static analyzer: `error()` masks NULL dereference

### ISSUE THREAD

**Issue by 251** [llvm/llvm-project#56548]
For the following program:
```C
#include <error.h>
#include <stddef.h>

int main(int argc, char * argv[]) {
	int *p = NULL;

	if (argc > 2) {
		error(0, 1, "test: %s", "foo");
		int ret = *p;
		return ret;
	}
}                                                                                                                        
```
the null-deref is missed as soon as I add the call to `error()`. Given that the first argument to `error` is `0`, it does not call `exit` and the program crashes (`./a.out 1 2`). Does the analysis just give up here or is the model for `error` incorrect?

(Tested with **LLVM 14.0.6**.)

**Comment by haoNoQ** [comment]
Yes, looks like it's [overapproximated](https://github.com/llvm/llvm-project/blob/main/clang/lib/StaticAnalyzer/Checkers/NoReturnFunctionChecker.cpp#L60) to always act as noreturn function, thanks, great catch!

**Comment by martong** [comment]
First of all, `error()` is a GNU extension, not defined neither in POSIX nor in the C standard. Nevertheless, we already support many GNU extensions. From the man pages:
```
       If status has a nonzero value, then error() calls [exit(3)](https://man7.org/linux/man-pages/man3/exit.3.html) to
       terminate the program using the given value as the exit status;
       otherwise it returns after printing the error message.
```
@haoNoQ  Perhaps we should handle `error()` in the `StdLibraryFunctionsChecker` instead of the `NoReturnFunctionChecker`? We'd need a new `ValueConstraint` to express the `noreturn` branch.
